{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "import openai\n",
        "from typing import List\n",
        "from pydantic import BaseModel, Field\n",
        "import pickle\n",
        "import networkx as nx\n",
        "import ast\n",
        "import re\n",
        "import json\n",
        "import chromadb\n",
        "import cProfile\n",
        "import pstats\n",
        "\n",
        "from datasets import load_dataset\n",
        "from pprint import pprint\n",
        "\n",
        "from langchain_groq import ChatGroq\n",
        "from langchain_chroma import Chroma\n",
        "from langchain_openai import OpenAIEmbeddings\n",
        "from langchain_core.utils.function_calling import convert_pydantic_to_openai_function\n",
        "from langchain.agents import tool\n",
        "from langchain_openai import OpenAI\n",
        "from langchain_core.prompts import ChatPromptTemplate\n",
        "from langchain_openai import ChatOpenAI\n",
        "from langchain_community.callbacks import get_openai_callback\n",
        "\n",
        "from commit import update\n",
        "from utils.utils import serialize_dict_to_json, deserialize_json_to_dict\n",
        "from utils.chunk import SimpleFixedLengthChunker\n",
        "from utils.compress import get_skeleton\n",
        "\n",
        "from dotenv import load_dotenv, find_dotenv\n",
        "_ = load_dotenv(find_dotenv()) # read local .env file\n",
        "\n",
        "openai.api_key = os.environ['OPENAI_API_KEY']\n",
        "\n",
        "sflc = SimpleFixedLengthChunker()\n",
        "\n",
        "dataset = load_dataset(\"lahirum/SWE_Experimental\", split=\"train\")\n",
        "# filter = [0, 1, 2, 3, 4,5, 6, 7, 8, 9]\n",
        "# dataset = dataset.select(filter)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "id": "a4fa7cd9",
      "metadata": {},
      "outputs": [],
      "source": [
        "def load_graph(pickle_path):\n",
        "    \"\"\"Loads a NetworkX DiGraph from a pickle file.\"\"\"\n",
        "    with open(pickle_path, \"rb\") as f:\n",
        "        graph = pickle.load(f)\n",
        "    return graph"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {},
      "outputs": [],
      "source": [
        "import prompts\n",
        "import schema\n",
        "from langchain.output_parsers.openai_functions import JsonOutputFunctionsParser\n",
        "# from langchain_deepseek import ChatDeepSeek\n",
        "\n",
        "llm = ChatOpenAI(\n",
        "    model=\"gpt-4o-mini\",\n",
        "    temperature=0.1,\n",
        "    #max_retries=2,\n",
        ")\n",
        "\n",
        "# llm_deepseek = ChatDeepSeek(\n",
        "#     model=\"deepseek-reasoner\",\n",
        "#     temperature=0,\n",
        "#     max_tokens=None,\n",
        "#     timeout=None,\n",
        "#     max_retries=2,\n",
        "#     # other params...\n",
        "# )\n",
        "\n",
        "llm_large = ChatOpenAI(\n",
        "    model=\"gpt-4o\",\n",
        "    temperature=0,\n",
        "    max_retries=2,\n",
        ")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/tmp/ipykernel_2303471/3732032862.py:4: LangChainDeprecationWarning: The function `_convert_pydantic_to_openai_function` was deprecated in LangChain 0.1.16 and will be removed in 1.0. Use :meth:`~langchain_core.utils.function_calling.convert_to_openai_function()` instead.\n",
            "  functions=[convert_pydantic_to_openai_function(schema.SuspiciousComponentOutput)],\n"
          ]
        }
      ],
      "source": [
        "parser = JsonOutputFunctionsParser()\n",
        "\n",
        "model_extract = llm_large.bind(\n",
        "    functions=[convert_pydantic_to_openai_function(schema.SuspiciousComponentOutput)],\n",
        "    function_call=\"auto\",\n",
        ")\n",
        "extract_chain = prompts.prompt_extract | model_extract\n",
        "\n",
        "model_select = llm.bind(\n",
        "    functions=[convert_pydantic_to_openai_function(schema.FileSuspicionOutput)],\n",
        "    function_call=\"auto\",\n",
        ")\n",
        "select_chain = prompts.file_path_filter_prompt | model_select\n",
        "\n",
        "model_filter_list = llm.bind(\n",
        "    functions=[convert_pydantic_to_openai_function(schema.SuspiciousFilesOutputList)],\n",
        "    function_call=\"auto\",\n",
        ")\n",
        "filter_list_chain = prompts.get_suspicious_file_list_from_list_of_files_prompt | model_filter_list \n",
        "\n",
        "model_select_list = llm.bind(\n",
        "    functions=[convert_pydantic_to_openai_function(schema.SuspiciousFilesOutputList)],\n",
        "    function_call=\"auto\",\n",
        ")\n",
        "select_list_chain = prompts.suspicious_files_filter_list_usingclfn_prompt | model_select_list\n",
        "\n",
        "model_select_with_reason = llm.bind(\n",
        "    functions=[convert_pydantic_to_openai_function(schema.SuspiciousFileReasoningOutput)],\n",
        "    function_call=\"auto\",\n",
        ")\n",
        "select_with_reason_chain = prompts.suspicious_files_with_reason_prompt | model_select_with_reason\n",
        "\n",
        "model_select_directory = llm.bind(\n",
        "    functions=[convert_pydantic_to_openai_function(schema.SuspiciousDirectoryOutput)],\n",
        "    function_call=\"auto\",\n",
        ")\n",
        "select_directory_chain = prompts.suspicious_directory_prompt | model_select_directory\n",
        "\n",
        "generate_multiple_descriptions = prompts.prompt_embedding_retriver | llm\n",
        "\n",
        "# deep_reasoning_chain = prompts.deep_reasoning_prompt | llm_deepseek"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [],
      "source": [
        "# def start(inputs):\n",
        "#     problem_description = inputs['problem_description']\n",
        "#     name = inputs['name']\n",
        "#     graph = inputs['graph'] \n",
        "#     commit_id = inputs['commit_id']\n",
        "#     graph = inputs['graph'] \n",
        "\n",
        "#     update(name, commit_id)\n",
        "\n",
        "#     with get_openai_callback() as callback:\n",
        "#         result = extract_chain.invoke({\"problem_description\": problem_description})\n",
        "#         print(callback)\n",
        "\n",
        "#     result = result.additional_kwargs['function_call']['arguments']\n",
        "#     result = json.loads(result)\n",
        "#     result['name'] = name\n",
        "#     result['problem_description'] = problem_description\n",
        "#     result['graph'] = graph\n",
        "    \n",
        "#     return result"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [],
      "source": [
        "def embedding_retriever(inputs):\n",
        "    problem_description = inputs['problem_description']\n",
        "    name = inputs['name']\n",
        "    \n",
        "    embeddings = OpenAIEmbeddings(model=\"text-embedding-3-small\")\n",
        "\n",
        "    chroma_client = chromadb.PersistentClient(f\"chroma_db\")\n",
        "    collection = chroma_client.get_collection(name=f\"{name}_chroma_index\")\n",
        "\n",
        "    vector_store = Chroma(\n",
        "        client=chroma_client,\n",
        "        collection_name=f\"{name}_chroma_index\",\n",
        "        embedding_function=embeddings,\n",
        "    )\n",
        "\n",
        "    results = vector_store.similarity_search(problem_description, k=10,)\n",
        "    # results = vector_store.max_marginal_relevance_search(problem_description, k=10, lambda_mult=0.5)\n",
        "\n",
        "    file = deserialize_json_to_dict(\"django_file_ids.json\")\n",
        "    structure = {}\n",
        "    for result in results:\n",
        "        file_ids = file[result.metadata[\"filename\"]].split(\":\")\n",
        "        chunk_docs_of_file = vector_store.get_by_ids(file_ids)\n",
        "        structure[result.metadata['filename']] = get_skeleton(sflc.dechunk_docs(chunk_docs_of_file))\n",
        "        \n",
        "    with get_openai_callback() as callback:\n",
        "        answer = select_with_reason_chain.invoke({\"problem_description\": problem_description, \"file_structure\": structure})\n",
        "        print(callback)\n",
        "    answer = json.loads(answer.additional_kwargs['function_call']['arguments'])\n",
        "\n",
        "    return answer['suspicious_files']\n",
        "    "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "django\n",
            "DiGraph with 23714 nodes and 30175 edges\n",
            "Node 'django/django/apps/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/ar/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ar_DZ/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/az/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/bg/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/bn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/bs/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ca/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ckb/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/cs/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/cy/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/da/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/de/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/de_CH/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/el/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en_AU/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en_GB/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/eo/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/es/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/es_AR/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/es_CO/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/es_MX/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/es_NI/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/es_PR/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/et/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/eu/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/fa/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/fi/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/fr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ga/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/gd/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/gl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/he/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/hi/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/hr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/hu/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/id/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ig/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/is/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/it/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ja/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ka/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/km/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/kn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ko/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ky/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/lt/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/lv/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/mk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ml/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/mn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ms/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/ms/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nb/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pt/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pt_BR/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ro/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ru/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sq/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sr_Latn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sv/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ta/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/te/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/tg/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/th/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/tk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/tr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/uk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/uz/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/vi/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/zh_Hans/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/zh_Hant/formats.py' does not exist.\n",
            "Node 'django/django/contrib/admindocs/urls.py' does not exist.\n",
            "Node 'django/django/contrib/auth/migrations/0012_alter_user_first_name_max_length.py' does not exist.\n",
            "Node 'django/django/contrib/gis/db/backends/postgis/const.py' does not exist.\n",
            "Node 'django/django/contrib/gis/db/models/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/gis/forms/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/gis/sitemaps/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/postgres/expressions.py' does not exist.\n",
            "Node 'django/django/contrib/redirects/migrations/0002_alter_redirect_new_path_help_text.py' does not exist.\n",
            "Node 'django/django/contrib/sites/checks.py' does not exist.\n",
            "Node 'django/django/core/cache/backends/redis.py' does not exist.\n",
            "Node 'django/django/core/checks/async_checks.py' does not exist.\n",
            "Node 'django/django/core/checks/compatibility/django_4_0.py' does not exist.\n",
            "Node 'django/django/core/checks/files.py' does not exist.\n",
            "Node 'django/django/core/management/commands/optimizemigration.py' does not exist.\n",
            "Node 'django/django/core/serializers/jsonl.py' does not exist.\n",
            "Node 'django/django/db/backends/sqlite3/_functions.py' does not exist.\n",
            "Node 'django/django/db/models/fields/json.py' does not exist.\n",
            "Node 'django/django/http/__init__.py' does not exist.\n",
            "Node 'django/django/template/autoreload.py' does not exist.\n",
            "Node 'django/django/test/__init__.py' does not exist.\n",
            "Node 'django/django/utils/connection.py' does not exist.\n",
            "Node 'django/django/views/__init__.py' does not exist.\n",
            "Node 'django/django/views/decorators/common.py' does not exist.\n",
            "Node 'django/tests/admin_changelist/urls.py' does not exist.\n",
            "Node 'django/tests/admin_custom_urls/urls.py' does not exist.\n",
            "Node 'django/tests/admin_docs/namespace_urls.py' does not exist.\n",
            "Node 'django/tests/admin_docs/urls.py' does not exist.\n",
            "Node 'django/tests/admin_inlines/urls.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/complex_app/models/__init__.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/configured_settings_manage.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/custom_templates/project_template/.hidden/render.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/management/commands/suppress_base_options_command.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/simple_app/models.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/urls.py' does not exist.\n",
            "Node 'django/tests/admin_utils/urls.py' does not exist.\n",
            "Node 'django/tests/admin_views/test_nav_sidebar.py' does not exist.\n",
            "Node 'django/tests/admin_views/test_skip_link_to_content.py' does not exist.\n",
            "Node 'django/tests/admin_views/urls.py' does not exist.\n",
            "Node 'django/tests/admin_widgets/urls.py' does not exist.\n",
            "Node 'django/django/conf/locale/ckb/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/one_config_app/apps.py' does not exist.\n",
            "Node 'django/tests/apps/two_configs_app/apps.py' does not exist.\n",
            "Node 'django/tests/apps/two_configs_one_default_app/apps.py' does not exist.\n",
            "Node 'django/tests/apps/two_default_configs_app/apps.py' does not exist.\n",
            "Node 'django/tests/async_queryset/__init__.py' does not exist.\n",
            "Node 'django/tests/async_queryset/models.py' does not exist.\n",
            "Node 'django/tests/async_queryset/tests.py' does not exist.\n",
            "Node 'django/tests/auth_tests/backend_alias.py' does not exist.\n",
            "Node 'django/tests/auth_tests/models/__init__.py' does not exist.\n",
            "Node 'django/tests/auth_tests/models/with_unique_constraint.py' does not exist.\n",
            "Node 'django/tests/auth_tests/settings.py' does not exist.\n",
            "Node 'django/tests/auth_tests/urls_admin.py' does not exist.\n",
            "Node 'django/tests/backends/base/app_unmigrated/migrations/0001_initial.py' does not exist.\n",
            "Node 'django/tests/backends/base/app_unmigrated/models.py' does not exist.\n",
            "Node 'django/tests/backends/base/test_client.py' does not exist.\n",
            "Node 'django/tests/backends/base/test_introspection.py' does not exist.\n",
            "Node 'django/tests/backends/mysql/test_operations.py' does not exist.\n",
            "Node 'django/tests/backends/postgresql/test_operations.py' does not exist.\n",
            "Node 'django/tests/backends/sqlite/test_creation.py' does not exist.\n",
            "Node 'django/tests/backends/sqlite/test_features.py' does not exist.\n",
            "Node 'django/tests/backends/sqlite/test_functions.py' does not exist.\n",
            "Node 'django/tests/backends/sqlite/test_operations.py' does not exist.\n",
            "Node 'django/tests/builtin_server/urls.py' does not exist.\n",
            "Node 'django/tests/builtin_server/views.py' does not exist.\n",
            "Node 'django/tests/cache/tests_async.py' does not exist.\n",
            "Node 'django/tests/check_framework/apps.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/__init__.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/different_tags_app/__init__.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/different_tags_app/apps.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/different_tags_app/templatetags/__init__.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/different_tags_app/templatetags/different_tags.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_1/__init__.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_1/apps.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_1/templatetags/__init__.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_1/templatetags/same_tags.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_2/__init__.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_2/apps.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_2/templatetags/__init__.py' does not exist.\n",
            "Node 'django/tests/check_framework/template_test_apps/same_tags_app_2/templatetags/same_tags.py' does not exist.\n",
            "Node 'django/tests/check_framework/test_4_0_compatibility.py' does not exist.\n",
            "Node 'django/tests/check_framework/test_async_checks.py' does not exist.\n",
            "Node 'django/tests/check_framework/test_files.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/bad_class_based_error_handlers.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/bad_error_handlers_invalid_path.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/bad_function_based_error_handlers.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/beginning_with_slash.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/cbv_as_view.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/contains_tuple.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/good_class_based_error_handlers.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/good_function_based_error_handlers.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/include_contains_tuple.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/include_with_dollar.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/name_with_colon.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/no_warnings.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/no_warnings_i18n.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/non_unique_namespaces.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/path_compatibility/beginning_with_caret.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/path_compatibility/contains_re_named_group.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/path_compatibility/ending_with_dollar.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/unique_namespaces.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/warning_in_include.py' does not exist.\n",
            "Node 'django/tests/conditional_processing/urls.py' does not exist.\n",
            "Node 'django/tests/contenttypes_tests/test_migrations.py' does not exist.\n",
            "Node 'django/tests/contenttypes_tests/urls.py' does not exist.\n",
            "Node 'django/tests/context_processors/urls.py' does not exist.\n",
            "Node 'django/tests/csrf_tests/csrf_token_error_handler_urls.py' does not exist.\n",
            "Node 'django/tests/db_functions/comparison/test_collate.py' does not exist.\n",
            "Node 'django/tests/db_functions/comparison/test_json_object.py' does not exist.\n",
            "Node 'django/tests/dbshell/test_sqlite.py' does not exist.\n",
            "Node 'django/tests/dbshell/tests.py' does not exist.\n",
            "Node 'django/tests/deprecation/test_csrf_cookie_masked.py' does not exist.\n",
            "Node 'django/tests/deprecation/test_middleware_mixin.py' does not exist.\n",
            "Node 'django/tests/deprecation/test_use_l10n.py' does not exist.\n",
            "Node 'django/tests/file_storage/urls.py' does not exist.\n",
            "Node 'django/tests/file_uploads/urls.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/absolute_urls.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/no_slash_urls.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/settings.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/urls.py' does not exist.\n",
            "Node 'django/tests/foreign_object/models/__init__.py' does not exist.\n",
            "Node 'django/tests/forms_tests/field_tests/test_jsonfield.py' does not exist.\n",
            "Node 'django/tests/forms_tests/templatetags/__init__.py' does not exist.\n",
            "Node 'django/tests/forms_tests/templatetags/tags.py' does not exist.\n",
            "Node 'django/tests/forms_tests/tests/__init__.py' does not exist.\n",
            "Node 'django/tests/forms_tests/tests/test_deprecation_forms.py' does not exist.\n",
            "Node 'django/tests/forms_tests/urls.py' does not exist.\n",
            "Node 'django/tests/generic_inline_admin/urls.py' does not exist.\n",
            "Node 'django/tests/generic_views/urls.py' does not exist.\n",
            "Node 'django/tests/gis_tests/admin.py' does not exist.\n",
            "Node 'django/tests/gis_tests/gdal_tests/tests.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoadmin_deprecated/admin.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoadmin/urls.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoadmin_deprecated/__init__.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoadmin_deprecated/models.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoadmin_deprecated/tests.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoadmin_deprecated/urls.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoapp/sitemaps.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoapp/test_indexes.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoapp/urls.py' does not exist.\n",
            "Node 'django/tests/gis_tests/gis_migrations/migrations/0001_setup_extensions.py' does not exist.\n",
            "Node 'django/tests/gis_tests/gis_migrations/migrations/0002_create_models.py' does not exist.\n",
            "Node 'django/tests/gis_tests/rasterapp/migrations/0001_setup_extensions.py' does not exist.\n",
            "Node 'django/tests/gis_tests/rasterapp/migrations/0002_rastermodels.py' does not exist.\n",
            "Node 'django/tests/handlers/urls.py' does not exist.\n",
            "Node 'django/tests/i18n/commands/__init__.py' does not exist.\n",
            "Node 'django/tests/i18n/loading_app/apps.py' does not exist.\n",
            "Node 'django/tests/i18n/other/locale/fr/formats.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/default.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/disabled.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/included.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/namespace.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/path_unused.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/wrong.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/wrong_namespace.py' does not exist.\n",
            "Node 'django/tests/i18n/sampleproject/manage.py' does not exist.\n",
            "Node 'django/tests/i18n/unchanged/__init__.py' does not exist.\n",
            "Node 'django/tests/i18n/urls.py' does not exist.\n",
            "Node 'django/tests/i18n/urls_default_unprefixed.py' does not exist.\n",
            "Node 'django/tests/logging_tests/urls.py' does not exist.\n",
            "Node 'django/tests/logging_tests/urls_i18n.py' does not exist.\n",
            "Node 'django/tests/messages_tests/models.py' does not exist.\n",
            "Node 'django/tests/messages_tests/tests.py' does not exist.\n",
            "Node 'django/tests/middleware/cond_get_urls.py' does not exist.\n",
            "Node 'django/tests/middleware/extra_urls.py' does not exist.\n",
            "Node 'django/tests/middleware/urls.py' does not exist.\n",
            "Node 'django/tests/middleware_exceptions/urls.py' does not exist.\n",
            "Node 'django/tests/migrations/test_fake_initial_case_insensitive/fake_initial/0001_initial.py' does not exist.\n",
            "Node 'django/tests/migrations/test_fake_initial_case_insensitive/initial/0001_initial.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_conflict_long_name/0001_initial.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_conflict_long_name/0002_conflicting_second_migration_with_long_name.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_conflict_long_name/0002_second.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_manual_porting/0001_initial.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_manual_porting/0002_second.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_manual_porting/0003_third.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_manual_porting/0004_fourth.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_manual_porting/__init__.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_noop/0001_initial.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_noop/__init__.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_plan/0005_fifth.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_squashed_no_replaces/0001_squashed_0002.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_squashed_no_replaces/__init__.py' does not exist.\n",
            "Node 'django/tests/model_fields/test_jsonfield.py' does not exist.\n",
            "Node 'django/tests/model_meta/results.py' does not exist.\n",
            "Node 'django/tests/model_options/apps.py' does not exist.\n",
            "Node 'django/tests/model_options/test_default_pk.py' does not exist.\n",
            "Node 'django/tests/model_package/models/__init__.py' does not exist.\n",
            "Node 'django/tests/model_utils/__init__.py' does not exist.\n",
            "Node 'django/tests/model_utils/tests.py' does not exist.\n",
            "Node 'django/tests/postgres_tests/integration_settings.py' does not exist.\n",
            "Node 'django/tests/project_template/urls.py' does not exist.\n",
            "Node 'django/tests/proxy_models/admin.py' does not exist.\n",
            "Node 'django/tests/proxy_models/urls.py' does not exist.\n",
            "Node 'django/tests/queries/test_contains.py' does not exist.\n",
            "Node 'django/tests/queries/test_sqlcompiler.py' does not exist.\n",
            "Node 'django/tests/redirects_tests/urls.py' does not exist.\n",
            "Node 'django/tests/requests/test_accept_header.py' does not exist.\n",
            "Node 'django/tests/serializers/test_jsonl.py' does not exist.\n",
            "Node 'django/tests/servers/urls.py' does not exist.\n",
            "Node 'django/tests/sessions_tests/no_clear_expired.py' does not exist.\n",
            "Node 'django/tests/shortcuts/urls.py' does not exist.\n",
            "Node 'django/tests/sitemaps_tests/urls/index_only.py' does not exist.\n",
            "Node 'django/tests/sitemaps_tests/urls/sitemap_only.py' does not exist.\n",
            "Node 'django/tests/staticfiles_tests/settings.py' does not exist.\n",
            "Node 'django/tests/staticfiles_tests/test_handlers.py' does not exist.\n",
            "Node 'django/tests/staticfiles_tests/urls/default.py' does not exist.\n",
            "Node 'django/tests/syndication_tests/urls.py' does not exist.\n",
            "Node 'django/tests/template_tests/alternate_urls.py' does not exist.\n",
            "Node 'django/tests/template_tests/syntax_tests/i18n/test_blocktranslate.py' does not exist.\n",
            "Node 'django/tests/template_tests/syntax_tests/i18n/test_translate.py' does not exist.\n",
            "Node 'django/tests/template_tests/syntax_tests/test_debug.py' does not exist.\n",
            "Node 'django/tests/template_tests/test_autoreloader.py' does not exist.\n",
            "Node 'django/tests/template_tests/urls.py' does not exist.\n",
            "Node 'django/tests/test_client/urls.py' does not exist.\n",
            "Node 'django/tests/test_client/urls_middleware_urlconf.py' does not exist.\n",
            "Node 'django/tests/test_client_regress/urls.py' does not exist.\n",
            "Node 'django/tests/test_runner/test_shuffler.py' does not exist.\n",
            "Node 'django/tests/test_runner_apps/buffer/tests_buffer.py' does not exist.\n",
            "Node 'django/tests/test_runner_apps/failures/__init__.py' does not exist.\n",
            "Node 'django/tests/test_runner_apps/failures/tests_failures.py' does not exist.\n",
            "Node 'django/tests/test_runner_apps/tagged/tests_syntax_error.py' does not exist.\n",
            "Node 'django/tests/test_sqlite.py' does not exist.\n",
            "Node 'django/tests/test_utils/test_serializemixin.py' does not exist.\n",
            "Node 'django/tests/test_utils/test_simpletestcase.py' does not exist.\n",
            "Node 'django/tests/test_utils/urls.py' does not exist.\n",
            "Node 'django/tests/timezones/urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/converter_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/included_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/more_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/path_base64_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/path_dynamic_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/path_same_name_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/path_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/erroneous_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/extra_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/included_app_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/included_named_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/included_named_urls2.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/included_namespace_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/included_no_kwargs_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/included_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/included_urls2.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/named_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/named_urls_conflict.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/namespace_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/reverse_lazy_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/urlconf_outer.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/urls_error_handlers.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/urls_without_handlers.py' does not exist.\n",
            "Node 'django/tests/urlpatterns_reverse/views_broken.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/commands/mutually_exclusive_required_with_same_dest.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/commands/outputwrapper.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/commands/required_constant_option.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/commands/required_list_option.py' does not exist.\n",
            "Node 'django/tests/user_commands/urls.py' does not exist.\n",
            "Node 'django/tests/utils_tests/deconstructible_classes.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_connection.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_deconstruct.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_module/another_bad_module.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_module/another_good_module.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_module/bad_module.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_module/child_module/grandchild_module.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_module/good_module.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_module/main_module.py' does not exist.\n",
            "Node 'django/tests/validation/test_constraints.py' does not exist.\n",
            "Node 'django/tests/view_tests/default_urls.py' does not exist.\n",
            "Node 'django/tests/view_tests/generic_urls.py' does not exist.\n",
            "Node 'django/tests/view_tests/regression_21530_urls.py' does not exist.\n",
            "Node 'django/tests/view_tests/urls.py' does not exist.\n",
            "Node 'django/tests/xor_lookups/__init__.py' does not exist.\n",
            "Node 'django/tests/xor_lookups/models.py' does not exist.\n",
            "Node 'django/tests/xor_lookups/tests.py' does not exist.\n",
            "Checked out to 466920f6d726eee90d5566e0a9948e92b33a122e\n",
            "DiGraph with 25370 nodes and 32548 edges\n",
            "Tokens Used: 14680\n",
            "\tPrompt Tokens: 14460\n",
            "\t\tPrompt Tokens Cached: 0\n",
            "\tCompletion Tokens: 220\n",
            "\t\tReasoning Tokens: 0\n",
            "Successful Requests: 1\n",
            "Total Cost (USD): $0.002301\n",
            "django\n",
            "DiGraph with 25370 nodes and 32548 edges\n",
            "Node 'django/django/conf/locale/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/ar_DZ/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/az/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/cs/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/cy/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/de/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/de_CH/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/el/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en_AU/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en_GB/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/eo/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/fi/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/fr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/hr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/hu/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/id/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ig/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/ig/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/it/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ka/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ko/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ky/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/ky/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/lt/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/lv/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/mk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ml/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nb/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pt/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pt_BR/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ru/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sr_Latn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sv/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/tg/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/tg/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/tk/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/tk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/tr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/uk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/uz/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/uz/formats.py' does not exist.\n",
            "Node 'django/django/contrib/admindocs/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/auth/management/commands/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/auth/signals.py' does not exist.\n",
            "Node 'django/django/contrib/contenttypes/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/contenttypes/management/commands/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/flatpages/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/gis/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/gis/db/backends/postgis/const.py' does not exist.\n",
            "Node 'django/django/contrib/gis/management/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/gis/management/commands/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/humanize/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/messages/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/postgres/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/postgres/forms/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/redirects/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sessions/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sessions/management/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sessions/management/commands/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sitemaps/management/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sitemaps/management/commands/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sites/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/staticfiles/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/staticfiles/management/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/staticfiles/management/commands/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/syndication/__init__.py' does not exist.\n",
            "Node 'django/django/core/management/commands/__init__.py' does not exist.\n",
            "Node 'django/django/test/__init__.py' does not exist.\n",
            "Node 'django/setup.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/complex_app/management/__init__.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/complex_app/management/commands/__init__.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/management/__init__.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/management/commands/__init__.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/simple_app/management/__init__.py' does not exist.\n",
            "Node 'django/tests/admin_scripts/simple_app/management/commands/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/explicit_default_config_app/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/explicit_default_config_empty_apps/apps.py' does not exist.\n",
            "Node 'django/tests/apps/explicit_default_config_mismatch_app/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/no_config_app/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/one_config_app/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/two_configs_app/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/two_configs_one_default_app/__init__.py' does not exist.\n",
            "Node 'django/tests/apps/two_default_configs_app/__init__.py' does not exist.\n",
            "Node 'django/tests/asgi/__init__.py' does not exist.\n",
            "Node 'django/tests/async/__init__.py' does not exist.\n",
            "Node 'django/tests/auth_tests/models/__init__.py' does not exist.\n",
            "Node 'django/tests/auth_tests/settings.py' does not exist.\n",
            "Node 'django/tests/backends/base/app_unmigrated/__init__.py' does not exist.\n",
            "Node 'django/tests/backends/base/app_unmigrated/migrations/__init__.py' does not exist.\n",
            "Node 'django/tests/bash_completion/management/__init__.py' does not exist.\n",
            "Node 'django/tests/bash_completion/management/commands/__init__.py' does not exist.\n",
            "Node 'django/tests/builtin_server/urls.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/path_compatibility/contains_re_named_group.py' does not exist.\n",
            "Node 'django/tests/db_functions/math/test_random.py' does not exist.\n",
            "Node 'django/tests/dbshell/fake_client.py' does not exist.\n",
            "Node 'django/tests/empty_models/__init__.py' does not exist.\n",
            "Node 'django/tests/empty_models/models.py' does not exist.\n",
            "Node 'django/tests/file_uploads/urls.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/absolute_urls.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/no_slash_urls.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/urls.py' does not exist.\n",
            "Node 'django/tests/generic_views/urls.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoapp/urls.py' does not exist.\n",
            "Node 'django/django/conf/locale/ar_DZ/__init__.py' does not exist.\n",
            "Node 'django/tests/gis_tests/rasterapp/migrations/__init__.py' does not exist.\n",
            "Node 'django/tests/handlers/urls.py' does not exist.\n",
            "Node 'django/tests/i18n/loading_app/__init__.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/default.py' does not exist.\n",
            "Node 'django/tests/middleware/urls.py' does not exist.\n",
            "Node 'django/tests/middleware_exceptions/urls.py' does not exist.\n",
            "Node 'django/tests/migrations/test_fake_initial_case_insensitive/fake_initial/__init__.py' does not exist.\n",
            "Node 'django/tests/migrations/test_fake_initial_case_insensitive/initial/__init__.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_conflict_long_name/__init__.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_namespace_package/0001_initial.py' does not exist.\n",
            "Node 'django/tests/migrations/test_migrations_no_operations/__init__.py' does not exist.\n",
            "Node 'django/tests/model_enums/__init__.py' does not exist.\n",
            "Node 'django/tests/model_meta/results.py' does not exist.\n",
            "Node 'django/tests/staticfiles_tests/settings.py' does not exist.\n",
            "Node 'django/tests/syndication_tests/urls.py' does not exist.\n",
            "Node 'django/tests/test_client/urls.py' does not exist.\n",
            "Node 'django/tests/test_client_regress/urls.py' does not exist.\n",
            "Node 'django/tests/test_sqlite.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/path_same_name_urls.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/path_urls.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/__init__.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/commands/__init__.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/commands/no_system_checks.py' does not exist.\n",
            "Node 'django/tests/user_commands/management/commands/specific_system_checks.py' does not exist.\n",
            "Node 'django/tests/utils_tests/test_module/__main__.py' does not exist.\n",
            "Node 'django/tests/view_tests/urls.py' does not exist.\n",
            "Checked out to df46b329e0900e9e4dc1d60816c1dce6dfc1094e\n",
            "DiGraph with 23392 nodes and 29686 edges\n",
            "Tokens Used: 3338\n",
            "\tPrompt Tokens: 3112\n",
            "\t\tPrompt Tokens Cached: 0\n",
            "\tCompletion Tokens: 226\n",
            "\t\tReasoning Tokens: 0\n",
            "Successful Requests: 1\n",
            "Total Cost (USD): $0.0006023999999999999\n",
            "django\n",
            "DiGraph with 23392 nodes and 29686 edges\n",
            "Node 'django/django/bin/django-admin.py' does not exist.\n",
            "Node 'django/django/conf/locale/__init__.py' does not exist.\n",
            "Node 'django/tests/gis_tests/maps/__init__.py' does not exist.\n",
            "Node 'django/django/conf/locale/az/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/cs/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/cy/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/de/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/de_CH/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/el/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en_AU/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/en_GB/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/eo/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/fi/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/fr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/hr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/hu/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/id/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/it/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ka/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ko/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/lt/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/lv/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/mk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ml/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nb/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/nn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pt/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/pt_BR/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/ru/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sk/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sl/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sr_Latn/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/sv/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/tr/formats.py' does not exist.\n",
            "Node 'django/django/conf/locale/uk/formats.py' does not exist.\n",
            "Node 'django/django/contrib/admindocs/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/contenttypes/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/flatpages/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/gis/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/gis/db/backends/postgis/const.py' does not exist.\n",
            "Node 'django/django/contrib/humanize/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/postgres/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/postgres/forms/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/redirects/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sessions/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/sites/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/staticfiles/__init__.py' does not exist.\n",
            "Node 'django/django/contrib/syndication/__init__.py' does not exist.\n",
            "Node 'django/django/test/__init__.py' does not exist.\n",
            "Node 'django/tests/admin_inlines/urls.py' does not exist.\n",
            "Node 'django/tests/admin_views/urls.py' does not exist.\n",
            "Node 'django/tests/apps/default_config_app/__init__.py' does not exist.\n",
            "Node 'django/tests/auth_tests/models/__init__.py' does not exist.\n",
            "Node 'django/tests/auth_tests/settings.py' does not exist.\n",
            "Node 'django/tests/check_framework/urls/path_compatibility/contains_re_named_group.py' does not exist.\n",
            "Node 'django/tests/file_uploads/urls.py' does not exist.\n",
            "Node 'django/tests/flatpages_tests/urls.py' does not exist.\n",
            "Node 'django/tests/forms_tests/tests/__init__.py' does not exist.\n",
            "Node 'django/tests/generic_views/urls.py' does not exist.\n",
            "Node 'django/tests/gis_tests/admin.py' does not exist.\n",
            "Node 'django/tests/gis_tests/geoapp/urls.py' does not exist.\n",
            "Node 'django/tests/handlers/urls.py' does not exist.\n",
            "Node 'django/tests/i18n/patterns/urls/default.py' does not exist.\n",
            "Node 'django/tests/middleware/urls.py' does not exist.\n",
            "Node 'django/tests/middleware_exceptions/urls.py' does not exist.\n",
            "Node 'django/tests/model_meta/results.py' does not exist.\n",
            "Node 'django/tests/staticfiles_tests/settings.py' does not exist.\n",
            "Node 'django/tests/syndication_tests/urls.py' does not exist.\n",
            "Node 'django/tests/test_client/urls.py' does not exist.\n",
            "Node 'django/tests/test_client_regress/urls.py' does not exist.\n",
            "Node 'django/tests/test_sqlite.py' does not exist.\n",
            "Node 'django/tests/urlpatterns/path_urls.py' does not exist.\n",
            "Node 'django/tests/view_tests/urls.py' does not exist.\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "\n",
            "KeyboardInterrupt\n",
            "\n"
          ]
        }
      ],
      "source": [
        "retrivals = []\n",
        "\n",
        "for i in range(2, 30):\n",
        "    ips = {}\n",
        "\n",
        "    ips[\"commit_id\"] = dataset[i]['base_commit']\n",
        "    ips[\"name\"]= dataset[i]['instance_id'].split(\"__\")[0]\n",
        "    ips[\"problem_description\"] = dataset[i]['problem_statement']\n",
        "    ips[\"graph\"] = load_graph(f\"graph_{ips['name']}.pkl\")\n",
        "\n",
        "    answer = generate_multiple_descriptions.invoke({\"problem_description\": ips[\"problem_description\"]})\n",
        "\n",
        "    ips[\"problem_description\"] = \\\n",
        "        f\"\"\"## **Original GitHub issue description**:\\n\\n{ips[\"problem_description\"]}\\n\\n\\n## **Generated descriptions**:\\n\\n{answer.content}\"\"\"\n",
        "\n",
        "    update(ips['name'], ips['commit_id'])\n",
        "    ans = embedding_retriever(ips)\n",
        "\n",
        "    temp_dict = {dataset[i]['instance_id']: {}}\n",
        "    for level, a in enumerate(ans):\n",
        "        temp_dict[dataset[i]['instance_id']][level] = {\n",
        "            \"identified_file\": a['file'],\n",
        "            \"erroneous_file\": dataset[i][\"erroneous_file\"].strip(),\n",
        "        }\n",
        "    retrivals.append(temp_dict)\n",
        "    serialize_dict_to_json(retrivals, \"embedding_based_retrievals.json\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5e1130f7",
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
